File: typescript/src/ancilis/cli/watch-display.ts
/** ancilis.cli.watch-display — terminal rendering for watch mode. */

const USE_COLOR = process.env["NO_COLOR"] === undefined;

const RESET = USE_COLOR ? "\x1b[0m" : "";
const GREEN = USE_COLOR ? "\x1b[32m" : "";
const RED = USE_COLOR ? "\x1b[31m" : "";
const DIM = USE_COLOR ? "\x1b[2m" : "";
const BOLD = USE_COLOR ? "\x1b[1m" : "";

const STATUS_MARKS: Record<string, string> = {
  pass: "\u2713",
  fail: "\u2717",
  skip: "\u2013",
};
const STATUS_COLORS: Record<string, string> = {
  pass: GREEN,
  fail: RED,
  skip: DIM,
};

export interface WatchControlResult {
  id: string;
  name: string;
  status: "pass" | "fail" | "skip";
  evaluations: number;
  failures: number;
  flags: number;
}

export function formatHeader(agentName: string, posture: string, totalEvals: number): string {
  const now = new Date();
  const ts = now.toTimeString().slice(0, 8);
  const postureColor = posture === "compliant" ? GREEN : RED;
  return `${DIM}${ts}${RESET} ${postureColor}${posture}${RESET} \u2014 ${agentName} (${totalEvals} evals)`;
}

export function formatDelta(
  prevResults: WatchControlResult[] | null,
  newResults: WatchControlResult[],
): string[] {
  if (prevResults === null) return [];
  const prevMap = new Map(prevResults.map(r => [r.id, r.status]));
  const lines: string[] = [];
  for (const ctrl of newResults) {
    const old = prevMap.get(ctrl.id);
    const next = ctrl.status;
    if (old !== undefined && old !== next) {
      const oldMark = STATUS_MARKS[old] ?? "?";
      const newMark = STATUS_MARKS[next] ?? "?";
      lines.push(`  ${ctrl.name}: ${oldMark} \u2192 ${newMark}`);
    }
  }
  return lines;
}

export function printScanResult(opts: {
  agentName: string;
  controlResults: WatchControlResult[];
  posture: string;
  totalEvals: number;
  prevResults: WatchControlResult[] | null;
  changedPaths: string[] | null;
  clear: boolean;
}): void {
  if (opts.clear) {
    process.stdout.write("\x1b[2J\x1b[H");
  }
  process.stdout.write(formatHeader(opts.agentName, opts.posture, opts.totalEvals) + "\n");

  if (opts.changedPaths && opts.changedPaths.length > 0) {
    const shown = opts.changedPaths.slice(0, 3);
    const suffix = opts.changedPaths.length > 3 ? "..." : "";
    process.stdout.write(`  ${DIM}Changed: ${shown.join(", ")}${suffix}${RESET}\n`);
  }

  const delta = formatDelta(opts.prevResults, opts.controlResults);
  if (delta.length > 0) {
    process.stdout.write(`  ${BOLD}Changes:${RESET}\n`);
    for (const line of delta) {
      process.stdout.write(line + "\n");
    }
  }

  for (const ctrl of opts.controlResults) {
    const mark = STATUS_MARKS[ctrl.status] ?? "?";
    const color = STATUS_COLORS[ctrl.status] ?? "";
    let detail = `${ctrl.evaluations} evals`;
    if (ctrl.failures > 0) detail += `, ${ctrl.failures} failures`;
    process.stdout.write(`  ${color}${mark}${RESET} ${color}${ctrl.name}${RESET} \u2014 ${ctrl.status} (${detail})\n`);
  }
}

export function printSessionSummary(
  startTime: Date,
  totalScans: number,
  finalResults: WatchControlResult[] | null,
  finalPosture: string | null,
): void {
  const elapsed = Date.now() - startTime.getTime();
  const minutes = Math.floor(elapsed / 60000);
  const seconds = Math.floor((elapsed % 60000) / 1000);
  process.stdout.write("\n");
  process.stdout.write(`${BOLD}Watch session ended${RESET} \u2014 ${minutes}m ${seconds}s, ${totalScans} scan(s)\n`);
  if (finalResults !== null && finalPosture !== null) {
    const color = finalPosture === "compliant" ? GREEN : RED;
    const passing = finalResults.filter(r => r.status === "pass").length;
    const total = finalResults.length;
    process.stdout.write(`  Final posture: ${color}${finalPosture}${RESET} \u2014 ${passing}/${total} controls passing\n`);
  }
}

File: python/src/ancilis/cli/watch_display.py
"""ancilis.cli.watch_display — terminal rendering for watch mode."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from rich.console import Console

console = Console(highlight=False)

_STATUS_MARKS = {
    "pass": "[green]\u2713[/]",
    "fail": "[red]\u2717[/]",
    "skip": "[dim]\u2013[/]",
}
_STATUS_COLORS = {"pass": "green", "fail": "red", "skip": "dim"}


def format_header(agent_name: str, posture: str, total_evals: int) -> str:
    color = "green" if posture == "compliant" else "red"
    ts = datetime.now().strftime("%H:%M:%S")
    return f"[dim]{ts}[/] [{color}]{posture}[/] \u2014 {agent_name} ({total_evals} evals)"


def format_delta(
    prev_results: list[dict[str, Any]] | None,
    new_results: list[dict[str, Any]],
) -> list[str]:
    """Return markup lines describing control status changes between scans."""
    if prev_results is None:
        return []
    prev_map = {r["id"]: r["status"] for r in prev_results}
    lines = []
    for ctrl in new_results:
        old = prev_map.get(ctrl["id"])
        new = ctrl["status"]
        if old is not None and old != new:
            old_mark = _STATUS_MARKS.get(old, "?")
            new_mark = _STATUS_MARKS.get(new, "?")
            lines.append(f"  {ctrl['name']}: {old_mark} \u2192 {new_mark}")
    return lines


def print_scan_result(
    agent_name: str,
    control_results: list[dict[str, Any]],
    posture: str,
    total_evals: int,
    prev_results: list[dict[str, Any]] | None,
    changed_paths: list[str] | None = None,
    clear: bool = False,
) -> None:
    if clear:
        console.clear()
    console.print(format_header(agent_name, posture, total_evals))
    if changed_paths:
        shown = changed_paths[:3]
        suffix = "..." if len(changed_paths) > 3 else ""
        console.print(f"  [dim]Changed: {', '.join(shown)}{suffix}[/]")
    delta = format_delta(prev_results, control_results)
    if delta:
        console.print("  [bold]Changes:[/]")
        for line in delta:
            console.print(line)
    for ctrl in control_results:
        mark = _STATUS_MARKS.get(ctrl["status"], "?")
        color = _STATUS_COLORS.get(ctrl["status"], "white")
        detail = f"{ctrl['evaluations']} evals"
        if ctrl["failures"] > 0:
            detail += f", {ctrl['failures']} failures"
        console.print(f"  {mark} [{color}]{ctrl['name']}[/] \u2014 {ctrl['status']} ({detail})")


def print_session_summary(
    start_time: datetime,
    total_scans: int,
    final_results: list[dict[str, Any]] | None,
    final_posture: str | None,
) -> None:
    elapsed = datetime.now() - start_time
    minutes = int(elapsed.total_seconds()) // 60
    seconds = int(elapsed.total_seconds()) % 60
    console.print()
    console.print(f"[bold]Watch session ended[/] \u2014 {minutes}m {seconds}s, {total_scans} scan(s)")
    if final_results and final_posture:
        color = "green" if final_posture == "compliant" else "red"
        passing = sum(1 for r in final_results if r["status"] == "pass")
        total = len(final_results)
        console.print(
            f"  Final posture: [{color}]{final_posture}[/] \u2014 {passing}/{total} controls passing"
        )
